File: Red-Black Tree/RedBlackTree.java
class Node {
    private char key;
    private Node right;
    private Node left;
    private Node parent;
    private boolean red;

    public Node(char key) {
        this.key = key;
        this.red = true; // Node baru selalu berwarna merah
    }

    public void setRight(Node right) {
        this.right = right;
    }

    public void setLeft(Node left) {
        this.left = left;
    }

    public void setParent(Node parent) {
        this.parent = parent;
    }

    public void setRed(boolean red) {
        this.red = red;
    }

    public Node getRight() {
        return right;
    }

    public Node getLeft() {
        return left;
    }

    public Node getParent() {
        return parent;
    }

    public char getKey() {
        return key;
    }

    public boolean isRed() {
        return red;
    }
}

class Tree {
    private Node root;

    public Tree() {
        this.root = null;
    }

    public void add(char key) {
        // Implementasi penambahan node ke dalam Red-Black Tree
        // ...
    }

    public boolean remove(char key) {
        // Implementasi penghapusan node dari Red-Black Tree
        // ...
        return false;
    }

    public boolean isExist(char key) {
        // Implementasi pengecekan keberadaan node dalam Red-Black Tree
        // ...
        return false;
    }

    private void rotateToRight(Node current) {
        // Implementasi rotasi ke kanan pada Red-Black Tree
        // ...
    }

    private void rotateToLeft(Node current) {
        // Implementasi rotasi ke kiri pada Red-Black Tree
        // ...
    }

    // Metode untuk menampilkan Tree dengan algoritma pre-order, in-order, dan post-order
    private void preOrder(Node node) {
        if (node != null) {
            System.out.print(node.getKey() + " ");
            preOrder(node.getLeft());
            preOrder(node.getRight());
        }
    }

    private void inOrder(Node node) {
        if (node != null) {
            inOrder(node.getLeft());
            System.out.print(node.getKey() + " ");
            inOrder(node.getRight());
        }
    }

    private void postOrder(Node node) {
        if (node != null) {
            postOrder(node.getLeft());
            postOrder(node.getRight());
            System.out.print(node.getKey() + " ");
        }
    }

    public void displayTree() {
        System.out.println("Pre-order: ");
        preOrder(root);
        System.out.println("\n\nIn-order: ");
        inOrder(root);
        System.out.println("\n\nPost-order: ");
        postOrder(root);
    }
}

File: Hashmap/RedBlackTree.py
class Node:
    def __init__(self, key, value, color, left=None, right=None):
        self.key = key
        self.value = value
        self.color = color
        self.left = left
        self.right = right

class RedBlackTree:
    def __init__(self):
        self.NIL = Node(None, None, "BLACK")
        self.root = self.NIL

    def insert(self, key, value):
        self.root = self._insert(self.root, key, value)
        self.root.color = "BLACK"

    def _insert(self, root, key, value):
        if root == self.NIL:
            return Node(key, value, "RED", self.NIL, self.NIL)

        if key < root.key:
            root.left = self._insert(root.left, key, value)
        elif key > root.key:
            root.right = self._insert(root.right, key, value)
        else:
            root.value = value

        if self._is_red(root.right) and not self._is_red(root.left):
            root = self._rotate_left(root)
        if self._is_red(root.left) and self._is_red(root.left.left):
            root = self._rotate_right(root)
        if self._is_red(root.left) and self._is_red(root.right):
            self._flip_colors(root)

        return root

    def search(self, key):
        return self._search(self.root, key)

    def _search(self, root, key):
        while root != self.NIL:
            if key < root.key:
                root = root.left
            elif key > root.key:
                root = root.right
            else:
                return root.value
        return None

    def inorder_traversal(self):
        result = []
        self._inorder_traversal(self.root, result)
        return result

    def _inorder_traversal(self, root, result):
        if root != self.NIL:
            self._inorder_traversal(root.left, result)
            result.append((root.key, root.value))
            self._inorder_traversal(root.right, result)

    def _is_red(self, node):
        return node.color == "RED"

    def _rotate_left(self, h):
        x = h.right
        h.right = x.left
        x.left = h
        x.color = h.color
        h.color = "RED"
        return x

    def _rotate_right(self, h):
        x = h.left
        h.left = x.right
        x.right = h
        x.color = h.color
        h.color = "RED"
        return x

    def _flip_colors(self, h):
        h.color = "RED"
        h.left.color = "BLACK"
        h.right.color = "BLACK"
